shared/storage/minio.py

```diff
@@ -143,85 +157,118 @@ def create_root_storage(self, bucket_name="archive", region="us-east-1"):
 
-        except MinioException:
-            raise
+        if reduced_redundancy:
+            headers["x-amz-storage-class"] = "REDUCED_REDUNDANCY"
+
+        # it's safe to do a BinaryIO cast here because we know that put_object only uses a function of the shape:
+        # read(self, size: int = -1, /) -> bytes
+        # GZipStreamReader implements this (we did it ourselves)
+        # ZstdCompressionReader implements read(): https://github.com/indygreg/python-zstandard/blob/12a80fac558820adf43e6f16206120685b9eb880/zstandard/__init__.pyi#L233C5-L233C49
+        # BytesIO implements read(): https://docs.python.org/3/library/io.html#io.BufferedReader.read
+        # IO[bytes] implements read(): https://github.com/python/cpython/blob/3.13/Lib/typing.py#L3502
+
+        return self.minio_client.put_object(
+            bucket_name,
+            path,
+            cast(BinaryIO, result),
+            -1,
+            metadata=headers,
+            content_type="text/plain",
+            part_size=PART_SIZE,
+        )
 
     @overload
-    def read_file(self, bucket_name: str, path: str) -> bytes: ...
+    def read_file(
+        self, bucket_name: str, path: str, file_obj: None = None
+    ) -> bytes: ...
 
     @overload
-    def read_file(self, bucket_name: str, path: str, file_obj: BinaryIO) -> None: ...
+    def read_file(self, bucket_name: str, path: str, file_obj: IO[bytes]) -> None: ...
 
-    def read_file(self, bucket_name, path, file_obj=None) -> bytes | None:
+    def read_file(
+        self, bucket_name: str, path: str, file_obj: IO[bytes] | None = None
+    ) -> bytes | None:
+        headers: dict[str, str | list[str] | Tuple[str]] = {
+            "Accept-Encoding": "gzip, zstd"
+        }
         try:
-            res = self.minio_client.get_object(bucket_name, path)
-            if file_obj is None:
-                data = BytesIO()
-                for d in res.stream(CHUNK_SIZE):
-                    data.write(d)
-                data.seek(0)
-                return data.getvalue()
-            else:
-                for d in res.stream(CHUNK_SIZE):
-                    file_obj.write(d)
+            response = cast(
+                HTTPResponse,
+                self.minio_client.get_object(  # this returns an HTTPResponse
+                    bucket_name, path, request_headers=headers
+                ),
+            )
         except S3Error as e:
             if e.code == "NoSuchKey":
                 raise FileNotInStorageError(
                     f"File {path} does not exist in {bucket_name}"
                 )
             raise e
-        except MinioException:
-            raise
+        if response.headers:
+            content_encoding = response.headers.get("Content-Encoding", None)
+            if content_encoding == "zstd":
+                # we have to manually decompress zstandard compressed data
+                cctx = zstandard.ZstdDecompressor()
```

Comment:
Another thing I looked up yesterday:
Since this is builtin with `urllib3 > 2`, you could check `urlllib3.__version__` to make sure we only manually decompress with version 1.
Or maybe `urllib3` has a different flag to check whether it handles zstd internally?

Otherwise we end up in a situation where we update the dependency and then we try to decompress uncompressed bytes and then fail.